keras_cv/losses/focal.py

```diff
@@ -76,14 +76,16 @@ def call(self, y_true, y_pred):
         y_true = tf.cast(y_true, y_pred.dtype)
 
         if self.label_smoothing:
-            y_true = self._smooth_labels(y_true)
+            y_smoothed = self._smooth_labels(y_true)
+        else:
+            y_smoothed = y_true
```

Comment:
can we remove y_smoothed and just move smoothing to after we compute alpha?

Reply:
This implementation comes from breaking down another implementation which include possible fixes for #681

In some case, you still want to leave untouched the `y_true` variable, in order to perform other computation at the reduction steps ( see FSAF's implementation of focal loss here https://github.com/xuannianz/FSAF/blob/791ca41187eec25918b2626322c43c9ab7c06f61/losses.py#L22-L60 ). Therefore you want two variable, one for computing cross-entropy, possibly smoothed, and the original `y_true` variable.

With a night to think over, I wonder if the label smoothing should maybe just be left out of this focal loss computation. I think that there is a reason why pytorch uses a formulation with mulitplications instead of a conditional statement. In that case the formula extends to the [0,1] domain. Therefore the label smoothing can be removed from the loss, and implemented outside it. Maybe the solution to clarify label smoothing, would be finally to remove it from the `FocalLoss` all together.

What do you think @LukeWood @quantumalaviya ?

Reply:
Sorry, I just meant in terms of code structure:

```
    def call(self, y_true, y_pred):
        y_pred = tf.convert_to_tensor(y_pred)
        y_true = tf.cast(y_true, y_pred.dtype)

        if self.from_logits:
            y_pred = tf.nn.sigmoid(y_pred)

        alpha = self._alpha * y_true + (1.0 - self._alpha) * (1.0 - y_true)
        pt = y_true * y_pred + (1.0 - y_true) * (1.0 - y_pred)
        cross_entropy = K.binary_crossentropy(self._smooth_labels(y_true), y_pred)
        loss = alpha * tf.pow(1.0 - pt, self._gamma) * cross_entropy

        return tf.reduce_mean(loss, axis=-1)
```

If label_smoothing=0, its a no-op anyways so you can just call it directly.

Reply:
label smoothing makes sense for focal loss
Please go ahead and follow the CE loss implementation:
https://github.com/keras-team/keras/blob/b80dd12da9c0bc3f569eca3455e77762cf2ee8ef/keras/losses.py#L1978-L1988